reasoner/planner/AnswerCountEstimator.java

```diff
@@ -60,210 +59,204 @@
         }
     }
 
-    private Set<ResolvableConjunction> dependencies(Concludable concludable) {
-        return iterate(logicMgr.applicableRules(concludable).keySet()).map(rule -> rule.condition().conjunction()).toSet();
+    public long estimateAnswers(ResolvableConjunction conjunction, Set<Variable> variables) {
+        if (!conjunctionModels.containsKey(conjunction)) buildConjunctionModel(conjunction);
+        if (!fullConjunctionEstimators.containsKey(conjunction)) {
+            IncrementalEstimator incrementalEstimator = createIncrementalEstimator(conjunction);
+            conjunctionModels.get(conjunction).conjunctionSummary.resolvables().forEach(incrementalEstimator::extend);
+            fullConjunctionEstimators.put(conjunction, incrementalEstimator);
+        }
+        IncrementalEstimator estimator = fullConjunctionEstimators.get(conjunction);
+        Set<Variable> estimateableVariables = ReasonerPlanner.estimateableVariables(variables);
+        return estimator.answerEstimate(estimateableVariables);
     }
 
-    private static class ConjunctionConcludableStack {
-
-        ArrayList<ResolvableConjunction> stack;
-        Map<ResolvableConjunction, Concludable> concludables;
-
-        ConjunctionConcludableStack() {
-            this.stack = new ArrayList<>();
-            this.concludables = new HashMap<>();
-        }
+    public IncrementalEstimator createIncrementalEstimator(ResolvableConjunction conjunction) {
+        if (!conjunctionModels.containsKey(conjunction)) buildConjunctionModel(conjunction);
+        return new IncrementalEstimator(conjunctionModels.get(conjunction));
+    }
 
-        private void add(ResolvableConjunction conjunction) {
-            stack.add(conjunction);
-            concludables.put(conjunction, null);
-        }
+    public static class IncrementalEstimator {
+        private final ConjunctionModel conjunctionModel;
+        private final Map<LocalModel, Pair<Double, Optional<Variable>>> modelScale;
+        private final Map<Variable, Double> bestUnary;
+        private final Map<Variable, Set<LocalModel>> modelsWithVar;
 
-        private ResolvableConjunction last() {
-            return stack.get(stack.size() - 1);
+        private IncrementalEstimator(ConjunctionModel conjunctionModel) {
+            this.conjunctionModel = conjunctionModel;
+            this.modelScale = new HashMap<>();
+            this.bestUnary = new HashMap<>();
+            this.modelsWithVar = new HashMap<>();
         }
 
-        private boolean contains(ResolvableConjunction conjunction) {
-            return concludables.containsKey(conjunction);
+        public void extend(Resolvable<?> resolvable) {
+            Map<Variable, Double> unaryUpdates = new HashMap<>();
+            List<LocalModel> models = conjunctionModel.modelsForResolvable(resolvable);
+            for (LocalModel model : models) {
+                includeModelAndCheckUpdates(model, unaryUpdates);
```

Comment:
try to make the `includeModelAndCheckUpdates` do no changes to the class state, and treat it as a function that communicates with `propagate` purely by reading class state and updating the `unaryUpdates`.
Then, you can even have the function return a new `unaryUpdates` object rather than passing in an empty map and updating in C-style.